internal/common/tree.go

```diff
@@ -42,15 +43,53 @@ type DNSTreeNodeData struct {
 //   - Any node with the stopLabel and the propLabel:
 //   - Has the propLabel removed from itself and all it's children
 //   - Has the propLabel removed from any parent (or parent's parent) that has the label
-
 func PropagateStoppableLabel(node *DNSTreeNode, propLabel, value, stopLabel string) {
 	//propagate labels regardless of stop labels
-	propagateLabel(node, propLabel, value)
+	PropagateLabel(node, propLabel, value)
 
 	//propagate stop labels
 	resolveStops(node, propLabel, value, stopLabel)
 }
 
+func CopyLabel(label string, from, to *DNSTreeNode) {
+	toNode := FindNode(to, from.Name)
+	if toNode != nil {
+		// copy label values to any matching datasets
+		for _, fd := range from.Endpoints {
+			for _, td := range toNode.Endpoints {
+				if slices.Equal(fd.Targets, td.Targets) {
+					if fd.Labels == nil || fd.Labels[label] == "" {
+						if td.Labels != nil {
+							delete(td.Labels, label)
+						}
+					} else if fd.Labels[label] != "" {
+						if td.Labels == nil {
+							td.Labels = endpoint.NewLabels()
+						}
+						td.Labels[label] = fd.Labels[label]
+					}
+				}
+			}
+		}
+	}
+	for _, c := range from.Children {
+		CopyLabel(label, c, to)
+	}
+}
```

Comment:
Can you explain what this (`CopyLabel`) is doing? It looks like it is deciding equality of endpoints based on the targets, but not sure why you would ever do that.

Reply:
this is because there can be multiple endpoints with the same host (which is matched on line 55), but with different targets (i.e. GEO CNAMES), so the endpoint is only considered to be the same endpoint if it has the same host and targets (arguably this should also compare record TYPE, though that hasn't presented a problem yet).

Reply:
Targets can change, any difference in what desired wants the record targets to look like and what targets are already in the current will cause this to fail, they will never match.

An endpoint is unique within a record set when it's dnsName, Type and SetIdentifier are unique, this is what makes up it's [key](https://github.com/kubernetes-sigs/external-dns/blob/master/endpoint/endpoint.go#L306).

This is what you should be comparing when you are looking for endpoints that represent the same dns record in the provider, i.e. `fd.Key() == td.Key()`